### PR TITLE
WIP CRM-21115 messagetemplate api

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Support groupName/valueName parameters in MessageTemplate.get API as used throughout the codebase.  Fix PHP warnings when tokens are passed to the smarty processor.

Before
----------------------------------------
Direct SQL queries as API didn't support required parameters.

After
----------------------------------------
Single API function using basic_get to retrieve message templates

---

 * [CRM-21115: Improve MessageTemplate.Get API to support groupName\/valueName params](https://issues.civicrm.org/jira/browse/CRM-21115)

### CODE REVIEW

File: CRM/Core/BAO/MessageTemplate.php [hunk 1]
```diff
@@ -122,7 +122,7 @@ public static function add(&$params) {
   public static function del($messageTemplatesID) {
     // make sure messageTemplatesID is an integer
     if (!CRM_Utils_Rule::positiveInteger($messageTemplatesID)) {
-      CRM_Core_Error::fatal(ts('Invalid Message template'));
+      throw new Exception('Invalid Message template');
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/MessageTemplate.php]: it should be CRM_Core_Exception rather than just Exception

Comment (mattwire) [reply CRM/Core/BAO/MessageTemplate.php]: @eileenmcnaughton Cheers, I've replaced Exception with CRM_Core_Exception

**Comment by seamuslee001** [comment]
@mattwire test failure looks related

**Comment by mattwire** [comment]
@seamuslee001 :-( I'm a bit confused as to why that's failing. FWIW the same test fails with entity Order locally and I didn't touch that!  It's failing on the getcount API call which is using civicrm_api_basic_get internally.
With no params the correct count is returned:
```
Array
(
    [is_active] => 1
    [is_reserved] => 0
    [sequential] => 1
    [prettyprint] => 1
    [check_permissions] => 1
    [version] => 3
    [options] => Array
        (
            [is_count] => 1
        )

)
```

With MessageTemplateID IN 1 the correct count is returned:
```
Array
(
[sequential] => 1
[id] => Array
(
[IN] => Array
(
[0] => 1
)

)

[prettyprint] => 1
[check_permissions] => 1
[version] => 3
[options] => Array
(
[is_count] => 1
)

)
```
With MessageTemplateID NOT IN 1 it returns the wrong count:
```
Array
(
[sequential] => 1
[id] => Array
(
[NOT IN] => Array
(
[0] => 1
)

)

[prettyprint] => 1
[check_permissions] => 1
[version] => 3
[options] => Array
(
[is_count] => 1
)

)
```

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: api/v3/MessageTemplate.php [hunk 2]
```diff
@@ -80,6 +80,17 @@ function civicrm_api3_message_template_delete($params) {
 function _civicrm_api3_message_template_get_spec(&$params) {
   // fetch active records by default
   $params['is_active']['api.default'] = 1;
+  $params['is_reserved']['api.default'] = 0;
```
Comment (seamuslee001) [inline api/v3/MessageTemplate.php]: @mattwire doing some local testing it appears the addition of this line is breaking the syntax conformance test.

Comment (mattwire) [reply api/v3/MessageTemplate.php]: @seamuslee001  Ok, I've removed that line, let's see if it works now

**Comment by mattwire** [comment]
Jenkins test this please

**Comment by eileenmcnaughton** [comment]
@mattwire this one has been in WIP state for quite a long time - lets close it & when it is reviewable you can re-open. (I would probably do the function extraction first in it's own PR to get that merged since function extractions combined with other changes are hard to review & can cause PRs to stall)

**Pull request closed without merging**